### PR TITLE
feat(influx): Add support for InfluxDB v2

### LINKED ISSUE #1974: InfluxDB 2.0 API support
Looks like the 'direct' api for 1.x will not work with 2.0.
Thankfully the changes seem pretty small, for comparison [1.x write api](https://docs.influxdata.com/influxdb/v1.7/tools/api/#write-http-endpoint) and [2.x write api](https://v2.docs.influxdata.com/v2.0/api/#operation/PostWrite)

The body remains the same (line protocol), only major change I believe is for auth (though user + password is probably also possible), instead of `db` query param there is `org` and `bucket`.
There is also an official [java client library](https://github.com/influxdata/influxdb-client-java), though I am not sure if micrometer needs all the functionality it provides.

Sorry if there is already something in place to configure micrometer so it works with influxdb 2.0 (directly), but I did not find it.

### PR DESCRIPTION
**Pull request by bednar**
Closes #1974
Closes #1423

Added support for InfluxDB v2. 

This change introduces new parameters for `Influx` registry:

- `org` - Specifies the destination organization for writes.
  - only for InfluxDB v2
- `bucket` - Specifies the destination bucket for writes.
  - only for InfluxDB v2
  - `bucket` is alias for `db` => user can specify either `db` nor `bucket`
- `token` - Authentication token to authorize API requests.
  - support both version of InfluxDB v2

The changes of existing parameters:

- `retentionDuration` - For the InfluxDB v2 the value should be an amount of seconds (e.g. 3600) and it's used when `autoCreateDb` is enabled
- `autoCreateDb` - This also support the InfluxDB v2.

### CODE REVIEW

**Comment by bednar** [comment]
Hi @jkschneider,

What is an ideal way to parse JSON response in a registry? Could I use `json-path` or try implement it without additionally dependencies?

https://github.com/micrometer-metrics/micrometer/blob/c68525346b9d6f3b8758b1f43b3b4120b199ca95/implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java#L347

https://github.com/micrometer-metrics/micrometer/blob/c68525346b9d6f3b8758b1f43b3b4120b199ca95/implementations/micrometer-registry-influx/build.gradle#L5

Regards

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (DangerousDarlow) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:437]: InfluxDB supports variable timestamp precision. Looking at the Micrometer documentation this doesn't seem to be an option?

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:437]: Yeah. Currently the Micrometer only supports millisecond precision for timestamp. See - https://github.com/micrometer-metrics/micrometer/blob/07701bd311af40419964a54add6ae60b3cfdf29b/micrometer-core/src/main/java/io/micrometer/core/instrument/Clock.java#L47

**Comment by bednar** [comment]
Hi @shakuzen, @jkschneider from my POV the PR is prepared to review. Is there anything else what can I do?

Thanks for your review.

**Comment by timhallinflux** [comment]
any progress on a review here?

**Comment by timhallinflux** [comment]
bump

**Comment by bednar** [comment]
Sources are rebased with master. 

Could we do something to review this PR?

**Comment by PiotrNowicki** [comment]
Guys, any updates on this or plans when Influx DB v2 will be supported?

**Comment by ivankudibal** [comment]
Hi Micrometer, is there any plan to merge this PR?

**Comment by fmmatthewzeemann** [comment]
Hi Micrometer, any updates on merging this PR? Really need  InfluxDB v2 support ...

**Comment by vertex-github** [comment]
+1 on this - Influxdb2 is up to v2.0.2 - Id really like to port our data storage to the new engine from our SpringBoot apps

**Comment by murphytalk** [comment]
+1 
hope to see the InfluxDbv2 solution gets merged sooner.

**Comment by timhallinflux** [comment]
bump

**Comment by pivotal-issuemaster** [comment]
@bednar Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2113)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2113) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@bednar Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2113)!

**Comment by overshareware** [comment]
Any timeline on this being merged? Desperate for v2 support.

**Comment by LimmaPaulus** [comment]
I have been using this PR with manual build for a long time on production. It has worked flawlessly.

**Comment by vertex-github** [comment]
Hey folks, is there something missing from the PR or an issue with the approach taken or is this just pending due to lack of resources for review / merge?

**Comment by shakuzen** [comment]
I've added the issue #1974 to the 1.7 milestone so we make sure to review this in that time frame. There's clearly a lot of interest in InfluxDB v2 support, so this will be prioritized.

**Comment by bednar** [comment]
@shakuzen thanks for awesome news. The PR is rebased and ready to review.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxConfig.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxConfig.java:148]:
```suggestion
     * @return The destination bucket (or db) for writes.
```

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxConfig.java]: Fixed

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:135]:
```suggestion
                    .post(config.uri() + "/api/v2/buckets");
```

Unfortunately, it seems the Influx v2 API returns a 200 response code even though the URI is wrong. We should add some tests with an InfluxDB instance running, using testcontainers, to ensure things are working as expected now and in the future. That doesn't need to block this pull request; I can follow-up with that work after merging this.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java:27]: It seems we need to generate a valid orgID, which is different from the org name. Passing an org name here results in the following error being returned:

```
{
	"code": "invalid",
	"message": "failed to unmarshal json: id must have a length of 16 bytes"
}
```

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java:27]: We have to check if the `org` parameter is `ID` otherwise retrieve `ID` by something like: [retrieveOrg](https://github.com/micrometer-metrics/micrometer/blob/229dcc3f4f2ab79e04fc85e1c22379323ee589cf/implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java#L338)
or It is a little bit out of scope of and we can remove the support for creating Bucket.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: Fixed `bucket` -> `buckets`. Thanks.

I will prepare an integration test over https://www.testcontainers.org/modules/databases/influxdb/.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java:27]: > It is a little bit out of scope of and we can remove the support for creating Bucket.

I'm leaning towards this direction. Perhaps we can revisit that option later. For initial v2 support, I think it is fine to not have it. Most companies will want to configure a retention policy on the bucket when creating it, and they will make tokens with specific access, so it seems highly likely that the token configured in the app won't have permission to create buckets anyways.

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java:27]: Ok, I will remove support for creating Bucket.

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/CreateBucketJSONBuilder.java]: The support for creating `Bucket` is removed.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: The support for creating `Bucket` is removed. 

Do you want to create integration tests over `testcontainers` for publishing metrics?

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:303]: As far as I can tell, only an admin or all access token can read the orgs. A read/write token (which is what I would expect to be provided for applications to write metrics) will not get a response with any orgs, even the one it belongs to. Here is what I get back with a read/write token:

```
{
    "links": {
        "self": "/api/v2/orgs"
    },
    "orgs": []
}
```

While it would be nice for users to have one less thing to configure, I am not sure it is worth having this attempt to retrieve the org. It also adds the JSON parser dependency. Users will have to configure the bucket and token anyways, so it isn't like we can provide something that works with defaults alone.

What do you think about removing the retrieve org logic?

File: micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java:294]: Why not `Collections.emptyMap()`?

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:283]: What exception are we anticipating that would end up in this catch block? We are defaulting to v2 above if there is an error calling the ping endpoint, so why is this defaulting to v1?

**Review by shakuzen — CHANGES_REQUESTED** [review]
We should note that userName and password are only applicable to InfluxDB 1.x. I've left some other comments in the code.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: I've created #2504 to track adding the integration tests. I don't want to block merging this enhancement on that. We want to include support for InfluxDB 2 in the 1.7 milestone release tomorrow if possible.

Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: Ok

File: micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java
Comment (bednar) [reply micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java]: Fixed

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:283]: > What exception are we anticipating that would end up in this catch block? 

Defaulting version to `v1`.

> We are defaulting to v2 above if there is an error calling the ping endpoint, so why is this defaulting to v1?

This implementation was prepared for early version of InfluxDB v2 which didn't supports the `/ping` endpoint. For the current version it doesn't sense => I changed the implementation. Now the the default version is set to `v1`.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: > What do you think about removing the retrieve org logic?

I agree. 

It is same case as a creating a `bucket` => the support for retrieve `org` is removed.

**Comment by bednar** [comment]
> We should note that userName and password are only applicable to InfluxDB 1.x. I've left some other comments in the code.

I updated docs for `userName` and `password` with: _The authenticating by 'userName' and 'password' is not supported for the InfluxDB v2._

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:280]: A problem with only determining the version from the results of a ping endpoint is that if the ping request happens to fail on the first publish, then the version will be set to v1 for the application, even if you wanted v2 and configured things for v2 and the influxdb backend is available after the ping request.

Since we aren't trying to retrieve the org anymore, perhaps we should add a config for `apiVersion` and have it default to v2 if `org` is configured, v1 otherwise. This way users can explicitly set the influx API version they intend to target.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [inline implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:120]: While not harmful (the write request is still successful with this), `rp` is not a parameter that the v2 write endpoint recognizes.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: Changed. The determining of InfluxDB version is based on the config.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (bednar) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java]: Fixed

File: micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java:316]: Since this isn't used anymore, we can get rid of these changes until we have a use for them, I think. If there is some use outside of supporting InfluxDB v2, we can consider them in a separate pull request.

**Review by shakuzen — APPROVED** [review]
Thank you very much @bednar for taking care of all the review so quickly and for the work in the first place. It is very much appreciated. I think this is almost ready to be merged. I will run a few more tests with it after I get some sleep, and then I can do any polishing along with merging it.

File: micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java
Comment (bednar) [reply micrometer-core/src/main/java/io/micrometer/core/ipc/http/HttpSender.java]: Fixed

File: implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java:49]: We should add tests that test `Basic` and `Token`, right now only `Bearer` auth is tested.

File: implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java:97]: I would use a mock instead of implementing it.

**Comment by jonatan-ivanov** [comment]
I talked to @shakuzen, let's merge this now and polish in a follow-up PR (see my comments).

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:91]: I was curious about this too in my testing. It will overwrite the `Authorization` header, which is the behavior we want, but we should make sure we are testing for it, to prevent any bugs from that behavior potentially changing in the future. I will add a note to the JavaDoc for the `userName` and `password` config methods to mention they won't be used if a token is configured.

> I would do this login this class not in the config.

Could you clarify what you mean?

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxApiVersion.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxApiVersion.java:39]: I suppose the problem is that if the `token` isn't configured, then a header shouldn't be added. That logic would need to be implemented at the callsite instead of here in that case.

File: implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java:97]: I don't think we mock config interfaces in other tests, so this is more in line with the code base.

File: implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java
Comment (shakuzen) [reply implementations/micrometer-registry-influx/src/test/java/io/micrometer/influx/InfluxMeterRegistryVersionsTest.java:49]: There's a test for `Token` already called `writeToV2Token`. I'll add tests for basic auth.

File: implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java
Comment (jonatan-ivanov) [reply implementations/micrometer-registry-influx/src/main/java/io/micrometer/influx/InfluxMeterRegistry.java:91]: typo, I wanted to write this: I would do this logic in this class not in the config.

**Pull request merged**
